libshvcore/src/utils/clioptions.cpp

```diff
@@ -400,53 +404,62 @@ bool ConfigCLIOptions::loadConfigFile()
 	}
 	return true;
 }
-
-std::string ConfigCLIOptions::configFile()
+namespace {
+bool is_absolute_path(const std::string &path)
 {
-	std::string config = "config";
-	std::string conf_ext = ".conf";
-	std::string config_file;
-	if(isValueSet(config)) {
-		config_file = value(config).toString();
-		if(config_file.empty()) {
-			/// explicitly set empty config means DO NOT load config from any file
-			return std::string();
-		}
-	}
-	else {
-		config_file = applicationName() + conf_ext;
-	}
 #ifdef _WIN32
-	bool is_absolute_path = config_file.size() > 2 && config_file[1] == ':';
+	return path.size() > 2 && path[1] == ':';
 #else
-	bool is_absolute_path = !config_file.empty() && config_file[0] == '/';
+	return !path.empty() && path[0] == '/';
 #endif
-	if(!is_absolute_path) {
-		std::string config_dir = configDir();
-		if(config_dir.empty())
-			config_dir = applicationDir();
-		config_file = config_dir + '/' + config_file;
-	}
-	if(!String::endsWith(config_file, conf_ext)) {
-		std::ifstream f(config_file);
-		if(!f)
-			config_file += conf_ext;
-	}
-	return config_file;
+}
+}
+std::string ConfigCLIOptions::configFile()
+{
+	auto [abs_config_dir, abs_config_file] = absoluteConfigPaths(configDir(), config());
+	return abs_config_file;
 }
 
 std::string ConfigCLIOptions::effectiveConfigDir()
 {
-	if(!configDir().empty())
-		return configDir();
-	String cfg = config();
-	if(!cfg.empty() && cfg[0] == '/') {
-		auto ix = cfg.lastIndexOf('/');
-		if(ix != std::string::npos) {
-			return  cfg.mid(0, ix);
+	auto [abs_config_dir, abs_config_file] = absoluteConfigPaths(configDir(), config());
+	return abs_config_dir;
+}
+
+std::tuple<std::string, std::string> ConfigCLIOptions::absoluteConfigPaths(const std::string &config_dir, const std::string &config_file) const
+{
+	using shv::core::utils::joinPath;
+	static const auto conf_ext = ".conf"s;
+	auto cwd = std::filesystem::current_path().string();
+	auto default_config_name = applicationName() + conf_ext;
+	if(config_file.empty()) {
+		if(config_dir.empty())
+			return make_tuple(cwd, joinPath(cwd, default_config_name));
+		else if(is_absolute_path(config_dir))
+			return make_tuple(config_dir, joinPath(config_dir, default_config_name));
+		else
+			return make_tuple(joinPath(cwd, config_dir), joinPath(cwd, config_dir, default_config_name));
+	}
+	else if(is_absolute_path(config_file)) {
+		if(config_dir.empty()) {
+			auto sep_pos = config_file.find_last_of('/');
+			// absolute config file must contain '/'
+			assert(sep_pos != std::string::npos);
```

Comment:
You already check for `is_absolute_path(config_file)`, doesn't that imply that there's a slash at the beginning? Also what if `config_file == "/"`? That's not a valid file, can you add a test for that?

Reply:
Then config dir will be empty and config file will be `/`, we will probably get exception later on opening config file what is IMO sufficient.

The assert() is superfluous, I'll remove it.

Reply:
yes

Reply:
test for missing slash or for config == `/`

Reply:
For config == /

Reply:
and what should I return in this case `("", "")`?

Reply:
I guess if you're fine with "/" returning just "/", then "" should return ""

Reply:
I'm still in doubts about handling the special case like `/`, what about `//` or `.`. I thing that we should solve all or nothing.

Reply:
I think the correct solution is to error out early (or at least according to what I've read in multiple places on the internet, e.g. here https://softwareengineering.stackexchange.com/a/419918).

I'm not exactly sure where to do that. Ideally the `CLIOPTION_GETTER_SETTER` should have some sort of an option to supply a validation callback. Then we wouldn't need to think about what `ConfigCLIOptions::absoluteConfigPaths()` should return in invalid cases.

Reply:
Correct solution is to check that absolute-config-dir and absolute-config-file exist and readable, what is not in scope of function `ConfigCLIOptions::absoluteConfigPaths()`. 

I cannot make some good test in this function and I also do not like to add one particular test for '/', so I suggest catch invalid config file/dir later.

Reply:
I agree that this is not the scope of this function.

I think that the invalid config file/dir should be checked even before this function is called. But that's another discussion/patch/something. :)

I'm resolving this.